### ISSUE TITLE
Cookbooks with same name but different paths in different groups cannot be resolved.

### ISSUE THREAD

**Issue by mavcunha** [berkshelf/berkshelf#1401]
```
berks --version
3.2.3
```

Given the following project structure:
```
├── Berksfile
├── Berksfile.lock
├── cookbooks
│   ├── backend
│   │   ├── metadata.rb
│   │   └── recipes
│   │       └── default.rb
│   └── frontend
│       ├── metadata.rb
│       └── recipes
│           └── default.rb
├── cookbooks-custom
│   └── book-one
│       ├── metadata.rb
│       └── recipes
│           └── default.rb
└── cookbooks-standard
    └── book-one
        ├── metadata.rb
        └── recipes
            └── default.rb
```

Note that `book-one` is defined in both `cookbooks-custom` and `cookbooks-standard`. Now if we define a `Berksfile` including the two `book-one` cookbooks but segregate them by groups, like so:

```
source "https://supermarket.getchef.com"

group :frontend do
  cookbook "book-one", :path => 'cookbooks-standard/book-one'
  cookbook "frontend", :path => 'cookbooks/frontend'
end

group :backend do
  cookbook "book-one", :path => 'cookbooks-custom/book-one'
  cookbook "backend", :path => 'cookbooks/backend'
end
```

Both `backend` and `frontend` declare `depends 'book-one'` in its `metadata.rb`

Berkshelf won't resolve for one of the groups, it seems that it treats both dependencies as the same and when it applies the filtering by groups it ends up with no `book-one` to find.

First, let's verify the group `backend` with `berks  --debug --only backend verify`:

```
I, [2015-03-18T09:17:15.096631 #79284]  INFO -- : Reducing lockfile
D, [2015-03-18T09:17:15.096705 #79284] DEBUG -- : Current lockfile:
D, [2015-03-18T09:17:15.096720 #79284] DEBUG -- :
D, [2015-03-18T09:17:15.096754 #79284] DEBUG -- :   DEPENDENCIES
D, [2015-03-18T09:17:15.096805 #79284] DEBUG -- :
D, [2015-03-18T09:17:15.096816 #79284] DEBUG -- :   GRAPH
D, [2015-03-18T09:17:15.096826 #79284] DEBUG -- :
D, [2015-03-18T09:17:15.096836 #79284] DEBUG -- : Unlocking dependencies no longer in the Berksfile
D, [2015-03-18T09:17:15.096846 #79284] DEBUG -- : Removing transitive dependencies
D, [2015-03-18T09:17:15.096907 #79284] DEBUG -- :   Checking backend (>= 0.0.0)
D, [2015-03-18T09:17:15.096962 #79284] DEBUG -- :     Skipping (not graphed)
D, [2015-03-18T09:17:15.096984 #79284] DEBUG -- :   Checking book-one (>= 0.0.0)
D, [2015-03-18T09:17:15.096994 #79284] DEBUG -- :     Skipping (not graphed)
D, [2015-03-18T09:17:15.097012 #79284] DEBUG -- : New lockfile:
D, [2015-03-18T09:17:15.097046 #79284] DEBUG -- :
D, [2015-03-18T09:17:15.097069 #79284] DEBUG -- :   DEPENDENCIES
D, [2015-03-18T09:17:15.097089 #79284] DEBUG -- :
D, [2015-03-18T09:17:15.097099 #79284] DEBUG -- :   GRAPH
D, [2015-03-18T09:17:15.097145 #79284] DEBUG -- :
Resolving cookbook dependencies...
I, [2015-03-18T09:17:15.097246 #79284]  INFO -- : Checking if lockfile is trusted
D, [2015-03-18T09:17:15.097289 #79284] DEBUG -- : Checking backend (>= 0.0.0)
D, [2015-03-18T09:17:15.097304 #79284] DEBUG -- :   Not in lockfile - cannot be trusted!
I, [2015-03-18T09:17:15.097359 #79284]  INFO -- : Installing from universe
Fetching 'backend' from source at cookbooks/backend
Fetching 'book-one' from source at cookbooks-custom/book-one
D, [2015-03-18T09:17:15.104082 #79284] DEBUG -- :   Creating a resolver
Fetching cookbook index from https://supermarket.getchef.com...
D, [2015-03-18T09:17:18.004953 #79284] DEBUG -- : ==> parsing Chef response body as JSON
D, [2015-03-18T09:17:18.167376 #79284] DEBUG -- :   Adding explicit dependency on backend (0.1.0) '/Users/mvaltas/Projects/berksbug/cookbooks/backend'
D, [2015-03-18T09:17:18.169429 #79284] DEBUG -- :   Adding explicit dependency on book-one (0.1.0) '/Users/mvaltas/Projects/berksbug/cookbooks-custom/book-one'
D, [2015-03-18T09:17:18.169467 #79284] DEBUG -- :   Starting resolution...
I, [2015-03-18T09:17:18.566580 #79284]  INFO -- : Installing backend (0.1.0)
D, [2015-03-18T09:17:18.566648 #79284] DEBUG -- :   Already installed - skipping install
Using backend (0.1.0) from source at cookbooks/backend
I, [2015-03-18T09:17:18.566832 #79284]  INFO -- : Installing book-one (0.1.0)
D, [2015-03-18T09:17:18.566852 #79284] DEBUG -- :   Already installed - skipping install
Using book-one (0.1.0) from source at cookbooks-custom/book-one
D, [2015-03-18T09:17:18.567023 #79284] DEBUG -- :   Finished resolving, calculating locks
D, [2015-03-18T09:17:18.567100 #79284] DEBUG -- :   New locks
D, [2015-03-18T09:17:18.567160 #79284] DEBUG -- :     backend (0.1.0)
D, [2015-03-18T09:17:18.567173 #79284] DEBUG -- :     book-one (0.1.0)
```

And now let's verify the group `frontend`:

```
I, [2015-03-18T09:17:15.096631 #79284]  INFO -- : Reducing lockfile
D, [2015-03-18T09:17:15.096705 #79284] DEBUG -- : Current lockfile:
D, [2015-03-18T09:17:15.096720 #79284] DEBUG -- :
D, [2015-03-18T09:17:15.096754 #79284] DEBUG -- :   DEPENDENCIES
D, [2015-03-18T09:17:15.096805 #79284] DEBUG -- :
D, [2015-03-18T09:17:15.096816 #79284] DEBUG -- :   GRAPH
D, [2015-03-18T09:17:15.096826 #79284] DEBUG -- :
D, [2015-03-18T09:17:15.096836 #79284] DEBUG -- : Unlocking dependencies no longer in the Berksfile
D, [2015-03-18T09:17:15.096846 #79284] DEBUG -- : Removing transitive dependencies
D, [2015-03-18T09:17:15.096907 #79284] DEBUG -- :   Checking backend (>= 0.0.0)
D, [2015-03-18T09:17:15.096962 #79284] DEBUG -- :     Skipping (not graphed)
D, [2015-03-18T09:17:15.096984 #79284] DEBUG -- :   Checking book-one (>= 0.0.0)
D, [2015-03-18T09:17:15.096994 #79284] DEBUG -- :     Skipping (not graphed)
D, [2015-03-18T09:17:15.097012 #79284] DEBUG -- : New lockfile:
D, [2015-03-18T09:17:15.097046 #79284] DEBUG -- :
D, [2015-03-18T09:17:15.097069 #79284] DEBUG -- :   DEPENDENCIES
D, [2015-03-18T09:17:15.097089 #79284] DEBUG -- :
D, [2015-03-18T09:17:15.097099 #79284] DEBUG -- :   GRAPH
D, [2015-03-18T09:17:15.097145 #79284] DEBUG -- :
Resolving cookbook dependencies...
I, [2015-03-18T09:17:15.097246 #79284]  INFO -- : Checking if lockfile is trusted
D, [2015-03-18T09:17:15.097289 #79284] DEBUG -- : Checking backend (>= 0.0.0)
D, [2015-03-18T09:17:15.097304 #79284] DEBUG -- :   Not in lockfile - cannot be trusted!
I, [2015-03-18T09:17:15.097359 #79284]  INFO -- : Installing from universe
Fetching 'backend' from source at cookbooks/backend
Fetching 'book-one' from source at cookbooks-custom/book-one
D, [2015-03-18T09:17:15.104082 #79284] DEBUG -- :   Creating a resolver
Fetching cookbook index from https://supermarket.getchef.com...
D, [2015-03-18T09:17:18.004953 #79284] DEBUG -- : ==> parsing Chef response body as JSON
D, [2015-03-18T09:17:18.167376 #79284] DEBUG -- :   Adding explicit dependency on backend (0.1.0) '/Users/mvaltas/Projects/berksbug/cookbooks/backend'
D, [2015-03-18T09:17:18.169429 #79284] DEBUG -- :   Adding explicit dependency on book-one (0.1.0) '/Users/mvaltas/Projects/berksbug/cookbooks-custom/book-one'
D, [2015-03-18T09:17:18.169467 #79284] DEBUG -- :   Starting resolution...
D, [2015-03-18T09:17:18.213485 #79284] DEBUG -- : Using 'chefignore' at '/Users/mvaltas/.berkshelf/cookbooks/oracle-xe-0.1.0/chefignore'
I, [2015-03-18T09:17:18.566580 #79284]  INFO -- : Installing backend (0.1.0)
D, [2015-03-18T09:17:18.566648 #79284] DEBUG -- :   Already installed - skipping install
Using backend (0.1.0) from source at cookbooks/backend
I, [2015-03-18T09:17:18.566832 #79284]  INFO -- : Installing book-one (0.1.0)
D, [2015-03-18T09:17:18.566852 #79284] DEBUG -- :   Already installed - skipping install
Using book-one (0.1.0) from source at cookbooks-custom/book-one
D, [2015-03-18T09:17:18.567023 #79284] DEBUG -- :   Finished resolving, calculating locks
D, [2015-03-18T09:17:18.567100 #79284] DEBUG -- :   New locks
D, [2015-03-18T09:17:18.567160 #79284] DEBUG -- :     backend (0.1.0)
D, [2015-03-18T09:17:18.567173 #79284] DEBUG -- :     book-one (0.1.0)
bismuto:berksbug $ berks --debug --only frontend verify
I, [2015-03-18T09:18:34.774878 #79299]  INFO -- : Reducing lockfile
D, [2015-03-18T09:18:34.774959 #79299] DEBUG -- : Current lockfile:
D, [2015-03-18T09:18:34.774977 #79299] DEBUG -- :
D, [2015-03-18T09:18:34.775419 #79299] DEBUG -- :   DEPENDENCIES
D, [2015-03-18T09:18:34.775439 #79299] DEBUG -- :     backend
D, [2015-03-18T09:18:34.775454 #79299] DEBUG -- :       path: cookbooks/backend
D, [2015-03-18T09:18:34.775465 #79299] DEBUG -- :     book-one
D, [2015-03-18T09:18:34.775499 #79299] DEBUG -- :       path: cookbooks-custom/book-one
D, [2015-03-18T09:18:34.775536 #79299] DEBUG -- :
D, [2015-03-18T09:18:34.775546 #79299] DEBUG -- :   GRAPH
D, [2015-03-18T09:18:34.775555 #79299] DEBUG -- :     backend (0.1.0)
D, [2015-03-18T09:18:34.775573 #79299] DEBUG -- :       book-one (>= 0.0.0)
D, [2015-03-18T09:18:34.775591 #79299] DEBUG -- :     book-one (0.1.0)
D, [2015-03-18T09:18:34.775601 #79299] DEBUG -- :
D, [2015-03-18T09:18:34.775629 #79299] DEBUG -- : Unlocking dependencies no longer in the Berksfile
D, [2015-03-18T09:18:34.775659 #79299] DEBUG -- :   Checking backend (>= 0.0.0)
D, [2015-03-18T09:18:34.775738 #79299] DEBUG -- :     Unlocking backend
D, [2015-03-18T09:18:34.775955 #79299] DEBUG -- :   Checking book-one (>= 0.0.0)
D, [2015-03-18T09:18:34.775979 #79299] DEBUG -- :     Unlocking book-one
D, [2015-03-18T09:18:34.776020 #79299] DEBUG -- : Removing transitive dependencies
D, [2015-03-18T09:18:34.776043 #79299] DEBUG -- :   Checking frontend (>= 0.0.0)
D, [2015-03-18T09:18:34.776066 #79299] DEBUG -- :     Skipping (not graphed)
D, [2015-03-18T09:18:34.776076 #79299] DEBUG -- : New lockfile:
D, [2015-03-18T09:18:34.776085 #79299] DEBUG -- :
D, [2015-03-18T09:18:34.776131 #79299] DEBUG -- :   DEPENDENCIES
D, [2015-03-18T09:18:34.776141 #79299] DEBUG -- :
D, [2015-03-18T09:18:34.776151 #79299] DEBUG -- :   GRAPH
D, [2015-03-18T09:18:34.776162 #79299] DEBUG -- :
Resolving cookbook dependencies...
I, [2015-03-18T09:18:34.776226 #79299]  INFO -- : Checking if lockfile is trusted
D, [2015-03-18T09:18:34.776247 #79299] DEBUG -- : Checking frontend (>= 0.0.0)
D, [2015-03-18T09:18:34.776259 #79299] DEBUG -- :   Not in lockfile - cannot be trusted!
I, [2015-03-18T09:18:34.802921 #79299]  INFO -- : Installing from universe
Fetching 'frontend' from source at cookbooks/frontend
D, [2015-03-18T09:18:34.806972 #79299] DEBUG -- :   Creating a resolver
Fetching cookbook index from https://supermarket.getchef.com...
D, [2015-03-18T09:18:37.262159 #79299] DEBUG -- : ==> parsing Chef response body as JSON
D, [2015-03-18T09:18:37.427158 #79299] DEBUG -- :   Adding explicit dependency on frontend (0.1.0) '/Users/mvaltas/Projects/berksbug/cookbooks/frontend'
D, [2015-03-18T09:18:37.427312 #79299] DEBUG -- :   Starting resolution...
Unable to satisfy constraints on package book-one, which does not exist, due to solution constraint (frontend = 0.1.0). Solution constraints that may result in a constraint on book-one: [(frontend = 0.1.0) -> (book-one >= 0.0.0)]
Missing artifacts: book-one
Demand that cannot be met: (frontend = 0.1.0)
Unable to find a solution for demands: frontend (0.1.0)
	/opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/resolver.rb:83:in `rescue in resolve'
	/opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/resolver.rb:73:in `resolve'
	/opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/installer.rb:175:in `install_from_universe'
	/opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/installer.rb:39:in `run'
	/opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/berksfile.rb:374:in `install'
	/opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/cli.rb:143:in `install'
	/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/thor-0.19.1/lib/thor/command.rb:27:in `run'
	/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/thor-0.19.1/lib/thor/invocation.rb:126:in `invoke_command'
	/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/thor-0.19.1/lib/thor.rb:359:in `dispatch'
	/opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/cli.rb:52:in `dispatch'
	/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/thor-0.19.1/lib/thor/base.rb:440:in `start'
	/opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/cli.rb:27:in `execute!'
	/opt/chefdk/embedded/apps/berkshelf/bin/berks:5:in `<top (required)>'
	/usr/bin/berks:43:in `load'
	/usr/bin/berks:43:in `<main>'
```
 
I tried to debug but I didn't get too far, what I know is that the `Berksfile` is parsed correct and Dependencies will be added correctly together with its groups, but after that the `book-one` will be left with one entry which will belong to one of the groups (not sure if there's predictability over which group).

I also repeated this flow removing `Berksfile.lock` between verify calls so to avoid any more variables.

**Comment by patcon** [comment]
What's expected behaviour here?

You're aware that even when using multiple groups, only one dependency graph is resolved, right? (This is how dep resolution tools like rubygems work.) The groups are used to install a subset of packages, but there should only be one resolved graph.

**Comment by mavcunha** [comment]
The documentation states: `-o, [--only=one two three]    # Only cookbooks that are in these groups.` My expectation is that groups would be exclusively added to the dependency resolution algorithm. Which in the example would not consider the cookbook "book-one" defined in another group. 

In sum:
`berks --debug --only backend ` would use `book-one` from `cookbooks-custom` and ignore `cookbooks-standard` as it is defined in a non-included group. And
`berks --debug --only frondend` would use only `book-one` from `cookbooks-standard` cookbook for the dep resolution.

If a user calls from `berks --only frondend backend` that is a resolution conflict as there's two `book-one` available.

**Comment by patcon** [comment]
Unfortunately, judging by all the code comments and tests, and what I know about other dependency resolving tools, that is not the intended behaviour, and likely never will be :( Making it so would break expectations, ie. lockfiles don't contain multiple graphs, and if they did, they would be ambiguous, whereas the goal is to lock down dependencies regardless of flags.

- https://github.com/berkshelf/berkshelf/blob/b0980478d2bcd9f22f6e54537e1f170895c4cfd4/lib/berkshelf/berksfile.rb#L52-L54
- https://github.com/berkshelf/berkshelf/blob/b0980478d2bcd9f22f6e54537e1f170895c4cfd4/spec/unit/berkshelf/installer_spec.rb
- https://github.com/berkshelf/berkshelf/blob/master/features/commands/install.feature

Basically, groups were created for environments (testing, production, etc), and changing the packages (in this case cookbooks) between environments would be anti-thetical to the goals of the tool.

His website is down, but Yehuda Katz (of Bundler fame, on which groups are modelled) had a great post on the purpose:
https://webcache.googleusercontent.com/search?q=cache:OAcWwm3-PPkJ:yehudakatz.com/2010/05/09/the-how-and-why-of-bundler-groups/+&cd=1&hl=en&ct=clnk&gl=us&client=ubuntu

Anyhow, I think this can be closed, if you'd like to do the honours. Sorry that these tools don't work for your use-case though, as that's always a bummer

**Comment by mavcunha** [comment]
Thanks Patrick for the clarificarion.
